qubesadmin/tools/qvm_template.py

```diff
@@ -0,0 +1,409 @@
+            return f.readlines()[-1].strip()
+    return subprocess.check_output(['lsb_release', '-sr'],
+        encoding='UTF-8').strip()
+
+parser = argparse.ArgumentParser(description='Qubes Template Manager')
+parser.add_argument('operation', type=str) 
+parser.add_argument('templates', nargs='*')
+
+# qrexec related
+parser.add_argument('--repo-files', action='append',
+    default=['/etc/yum.repos.d/qubes-templates.repo'],
+    help='Specify files containing DNF repository configuration.')
+parser.add_argument('--updatevm', default='sys-firewall',
+    help='Specify VM to download updates from.')
+# DNF-related options
+parser.add_argument('--enablerepo', action='append',
+    help='Enable additional repositories.')
+parser.add_argument('--disablerepo', action='append',
+    help='Disable certain repositories.')
+parser.add_argument('--repoid', action='append',
+    help='Enable just specific repositories.')
+parser.add_argument('--releasever', default=qubes_release(),
+    help='Override distro release version.')
+parser.add_argument('--cachedir', default=CACHE_DIR,
+    help='Override cache directory.')
+# qvm-template install
+parser.add_argument('--nogpgcheck', action='store_true',
+    help='Disable signature checks.')
+parser.add_argument('--allow-pv', action='store_true',
+    help='Allow setting virt_mode to pv in configuration file.')
+# qvm-template download
+parser.add_argument('--downloaddir', default='.',
+    help='Override download directory.')
+# qvm-template list
+parser.add_argument('--all', action='store_true')
+parser.add_argument('--installed', action='store_true')
+parser.add_argument('--available', action='store_true')
+parser.add_argument('--extras', action='store_true')
+parser.add_argument('--upgrades', action='store_true')
+
+# NOTE: Verifying RPMs this way is prone to TOCTOU. This is okay for local
+# files, but may create problems if multiple instances of `qvm-template` are
+# downloading the same file, so a lock is needed in that case.
+def verify_rpm(path, nogpgcheck=False, transaction_set=None):
+    if transaction_set is None:
+        transaction_set = rpm.TransactionSet()
+    with open(path, 'rb') as f:
+        try:
+            hdr = transaction_set.hdrFromFdno(f)
+            if hdr[rpm.RPMTAG_SIGSIZE] is None \
+                    and hdr[rpm.RPMTAG_SIGPGP] is None \
+                    and hdr[rpm.RPMTAG_SIGGPG] is None:
+                return nogpgcheck
+        except rpm.error as e:
+            if str(e) == 'public key not trusted' \
+                    or str(e) == 'public key not available':
+                return nogpgcheck
+            return False
+    return True
+
+def get_package_hdr(path, transaction_set=None):
+    if transaction_set is None:
+        transaction_set = rpm.TransactionSet()
+    with open(path, 'rb') as f:
+        hdr = transaction_set.hdrFromFdno(f)
+        return hdr
+
+def extract_rpm(name, path, target):
+    with open(path, 'rb') as in_file:
```

Comment:
Why? You don't seem to use `in_file` here.

Reply:
This seems to be the remains of the rpm2archive version. Will remove it later.